sdk/identity/identity/src/client/msalClient.ts

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { CredentialUnavailable } from "./errors";
+import {
+  PublicClientApplication,
+  Configuration,
+  AuthorizationCodeRequest,
+  AuthenticationResult,
+  DeviceCodeRequest,
+  ConfidentialClientApplication,
+  ClientCredentialRequest,
+  TokenCacheContext,
+  NetworkRequestOptions,
+  NetworkResponse,
+  INetworkModule
+} from "@azure/msal-node";
+import axios, { AxiosRequestConfig } from "axios";
+
+import { IdentityClient, TokenCredentialOptions } from "./identityClient";
+import { AccessToken } from "@azure/core-http";
+import { credentialLogger } from "../util/logging";
+import { NodeAuthOptions } from "@azure/msal-node/dist/config/Configuration";
+
+const logger = credentialLogger("InteractiveBrowserCredential");
+
+async function createPersistence(
+  cachePath?: string
+): Promise<
+  | {
+      cachePlugin?: {
+        beforeCacheAccess: (tokenCacheContext: TokenCacheContext) => Promise<void>;
+        afterCacheAccess: (tokenCacheContext: TokenCacheContext) => Promise<void>;
+      };
+    }
+  | undefined
+> {
+  let msalExt: any;
+  try {
+    msalExt = require("@azure/msal-node-extensions");
+  } catch (er) {
+    console.log("Can't find @azure/msal-node-extensions");
+    msalExt = null;
+    return;
+  }
+
+  console.log("process platform:", process.platform);
+
+  // On Windows, uses a DPAPI encrypted file
+  if (process.platform === "win32") {
+    const filePersistence = await msalExt.FilePersistenceWithDataProtection.create(
+      cachePath,
+      msalExt.DataProtectionScope.LocalMachine
+    );
+
+    return {
+      cachePlugin: new msalExt.PersistenceCachePlugin(filePersistence)
+    };
+  }
+
+  // On Mac, uses keychain.
+  if (process.platform === "darwin") {
+    const keychainPersistence = await msalExt.KeychainPersistence.create(
+      cachePath,
+      "serviceName",
+      "accountName"
+    );
+
+    return {
+      cachePlugin: new msalExt.PersistenceCachePlugin(keychainPersistence)
+    };
+  }
+
+  // On Linux, uses  libsecret to store to secret service. Libsecret has to be installed.
+  if (process.platform === "linux") {
+    const libSecretPersistence = await msalExt.LibSecretPersistence.create(
+      cachePath,
+      "serviceName",
+      "accountName"
+    );
+
+    return {
+      cachePlugin: new msalExt.PersistenceCachePlugin(libSecretPersistence)
```

Comment:
libsecret secret storage will fail if the service is not available, for instance when in an SSH session, We need to handle this failure and fallback to plaintext file storage if the credential options allow.